### DISCUSSION TITLE
background noise and reverberation

### DISCUSSION THREAD

**Discussion by StuartIanNaylor** [dscripka/openWakeWord#4] [General]
You can not really train for reverberation as in terms of spectrogram it radically changes the 'image' you get depending on room.
It just creates much less distinct blurred spectrograms based on mic position and room shape.

Beamforming like (Time Difference of Arrival) is often used to create directionality and reject weaker reverb in the audio capture.
Likely BSS multi-mic as it uses a similar principle to extract a signal from a multi signal mix also has a deverberation effect.
I am not that sure if Google x2 mics on there nest is just a simple beamformer and then they use single channel targetted voice extraction.
Voice extraction to extract a known traget voice seems to have better results than filtering out potentially unknown noise.
The only ASR that I know that seems to cope with elements of noise and reverb is OPenAi's whisper but not sure if part of the model does filter.

You can only add noise to a certain level where the KW is predominant and like reverb it makes the KW less distinct, but likely you can add noise at high level to your dataset and then filter that dataset so you train in filter artefacts as the artefact results with some filters  can be orders of magnitude less than noise level and have reltively little effect on resultant spectrogram.

The beamforming and targetted voice extraction due to high load dsp has always been lacking whilst big data use custom silicon to provide.

**Comment by dscripka** [comment]
I agree that simulating reverberation isn't ideal and likely introduces some artifacts. I haven't done a rigorous ablation study to see how much training on reverberated data helps performance, but given that the generated synthetic speech has a uniformly near-field sound I wanted to introduce as much variety as I could.

Beamforming is also a great method for handling this type of distortion, but as you mention is very hardware specific. I do have a good quality beamforming microphone that I've used for some testing (https://antimatter.ai/acusis-s), but sadly it looks like they are no longer manufacturing it. One of my goals for openWakeWord is to try and be as hardware independent as possible, so if the model itself can be trained to handle at least some reverberation without relying on beamforming, that's a start.

Adding noise has a similar effect (hopefully), and makes the model more robust. Training on de-noised data is an interesting idea; it looks like you have done some testing with DeepFilterNet (https://github.com/Rikorose/DeepFilterNet), so that may be a good place to start?

**Reply by StuartIanNaylor** [reply]
In https://github.com/StuartIanNaylor/ProjectEars/tree/main/ds there is a 2 channel beamformer which was my 1st attempt at C/C++
Python is a great RAD / Research lang but for DSP it sucks harder than a whirpool.

I finnally got a realtime GCC-PHAT going with the margins of direction added to the ringbuffer and its working 100% fairly low Pi load.
I should try swapping out the FFT code to some code that is optimised for neon and also optimise that code as basically it was me head scratching ok C/C++ lets just see if I can hack portaudio into this and create a ring buffer that is realtime.

The beamformer you have is a great piece of kit prob one of the best but there is a crucial thing its missing as its setup as a conference microphone and beams to predominant noise but there is no linkage where the KW envelope can set direction for a command sentence and provide a huge improvement for audio capture for ASR.

https://github.com/Rikorose/DeepFilterNet in terms of quality is right up there with RtxVoice but its single thread and struggles on low end Pi's aka my fave of a Pi02W as it can not keep up with the audio stream. I never benched on a Pi4 but on a rk3588 my new fave SBC its 30% of a big core.
Running this way low SNR still works amazingly well and fills that gap in the opensource voice processing but really only fits an all-in-one or central audio processor precusor to ASR.
It fails on voice noise as the model accepts that as voice but 90% is filtered very effectively at much better SNR levels than https://github.com/breizhn/DTLN
This is prob why Google have preference for targetted voice extraction with VoiceFilterLite

The exception to this rule is OPenAi's whisper as any DSP leaves a fingerprint and would likely need retraining with any filters fingerprint in the dataset.

**Reply by dscripka** [reply]
Thanks for the link to the beamformer code! That's a great option for a 2-channel beamformer, and to your point being able to integrate it directly with any keyword or ASR method is critical. The Acusis microphone does work really well, but the stock beamforming does just focus on whatever is loudest in the room and there is minimal control of that behavior.

One thing I'm curious to see is just how much performance is possible without a full mic array and beamformer combination. That is, if you have a good combination of noise cancellation, speaker identification and separation, and a wakeword model robust to the artifacts that those methods introduce, it might be possible to get sufficient performance for many applications with just normal hardware. Though likely all that is not possible to run on small SBCs.

For DeepFilterNet, I think one good experiment would be to mix some test wakeword clips with noise/music at low SNRs and see if the pre-trained openWakeWord models perform better with DeepFilterNet as a pre-processing step. To your point, the models might need to be re-trained due to filter artifacts, but I'll try to do some tests to get a better idea.

**Reply by StuartIanNaylor** [reply]
I have already done this with deepfilter and for KW it worked amazingly well with huge SNR and this is what confused me about whisper as the oppisite was true, but expect simpler ASR models to work as did the KWS and not even need artefact training even though that will increase accuracy and is easy accomplished.

The 2 mic hats,  https://plugable.com/products/usb-audio or https://learn.adafruit.com/adafruit-i2s-mems-microphone-breakout/ and just add 2x I2s mics on a I2S port.
Most USB like the Acusis are really great bits of kit but fail in being able to sync with KWS on 'lock on' to a command sentence and are really boardroom mics and other mic arrays often don't have audio out so you suffer from clock drift synchronisation.
There are 4mic and above hats that use TDM that split a channel at lower frequency to inject channels but seems to be a huge prob on the Pi of sync as on try the respeaker for example return a random channel order on each record which is a chocolate kettle.

There is DTLN that really would benefit from a C/C++ implementation as with all honesty Python sucks at iteration of even modest fequency DSP its a great RAD/Research lang but for embedded DSP its a woefull choice.
We don't seem to have any C/C++ or RUST devs if any that could accomplish DSP bandwidth without the load of Python.

The beamformer will help with far-field as that is where reverb increases in magnitude and 2x mics forward facing can give really good results for that, even if its hard to distinguish with humans ears.
There is Speex AEC that attenuates a ref signal (audio out), DTLN AEC but quite high load in comparison, but yeah individually effect might be minimal but together they all crreate an incremental chain.

I have been trying to find a low load 2 mic BSS alg or model as really you only have to run 2x vad models on the x2 nMic streams and you could use ondevice training to overfit VAD to users voice as the only prob with BSS is you never know the channel order of the split.
VAD models are usually very lite but in essence this is simple targetted voice extraction and if you think about predominant domestic use spatially its usually media or another and the target voice so the 2 mic BSS fits majority use.
The Math and code behind BSS is prob beyond me as the GCC_Phat had me screaming whilst even using multiple quite well documented sources.

There is also the esp32-s3 which is basically an esp32 with vector instructions that allows it almost 10x ML performance of various layers than its basic esp32 brother.
If you have a look at the esp32-box-lite its a product demo containing all of espressifs audio libs and pretty useless as a working product as to fit all much has had to comphremise but rather than a complete smart-assistant on a micro-controller it has the ability to be more specific as a KWS and cast off elements and use that load.
Also esp32 can be used in pairs where KWS & say squeezelite can be fused as a unit.

On all KWS as its far more resilient than ASR it doesn't mean the processed audio stream is past upwards as it can do its own thing and pass the raw audio stream to a more capable upstream audio processor for ASR.

**Comment by StuartIanNaylor** [comment]
Specifically with the beamformer the summing part of Delay Sum has zero load as its just that the sum of the current sample and delayed sample to give directionality.
The GCC-Phat to find the delay between mics creates 99% of the load and I think its possible to create a 4 mic directional with the same 2 mic alg because the geometry is known, if a actual useful geometry was created than just looking the part.
Its actually a 4 mic rectangle that the below invensense app note will validate.

The rear 2 mics is merely the reverse of the front 2 in terms of directionality as the angle is calculated by the front 2 and with known geometry we are talking simple pythagoras than complex gcc-phat to create an endfire out of x2 broadsides with 99% of the load being only the front x2 mic GCC-Phat calc.
Its just the respeaker mics tendency to be random that I never bothered to explore that avenue, also the geometries seem purely to look the part that actually fit a DS scheme. Delay sum do act as 1st order high pass filters but that is just EQ that can be applied at a lower 16khz resampled freq likely by an LADSPA plugin.
Invensense give a great 101 to DS beamforming that I often quote
https://invensense.tdk.com/wp-content/uploads/2015/02/Microphone-Array-Beamforming.pdf

Also there is always an assumption mems are better which isn't true they are just very convenient for mobile hardware but 6mm unidirectional electrets due to rear ports have physical rear rejection but seem to be relagated to the BS of looking the part than being functional.
Mems have tighter tolerances which is true but its a bit pointless when we are not working at that level of granularity.
Also especially in 2mic they should be front facing and not planar which creates no rear rejection that front facing does.
There is an absolute absence of even simple audio processing and specification in the majority of what is on offer.

I have tried MVDR & GSC beamformers but the load is orders higher than a Delay sum where TDOA is the only load really.
